engines/kingdom/logic.cpp

```diff
@@ -0,0 +1,605 @@
+	_opCodes[740] = &Logic::GPL3_740;
+	_opCodes[741] = &Logic::GPL3_741;
+	_opCodes[760] = &Logic::GPL3_760;
+	_opCodes[761] = &Logic::GPL3_761;
+	_opCodes[780] = &Logic::GPL3_780;
+	_opCodes[781] = &Logic::GPL3_781;
+	_opCodes[790] = &Logic::GPL3_790;
+	_opCodes[791] = &Logic::GPL3_791;
+
+	_opCodes[992] = &Logic::GPL4_992;
+	_opCodes[993] = &Logic::GPL4_993;
+}
+
+void Logic::enAll() {
+	_help = true;
+	_eye = true;
+	_replay = true;
+	_pouch = true;
+	_fstFwd = true;
+}
+
+void Logic::dsAll() {
+	_help = false;
+	_eye = false;
+	_replay = false;
+	_pouch = false;
+	_fstFwd = false;
+}
+
+void Logic::inventoryDel(int item) {
+	if (_inventory[item] > 0)
+		_inventory[item]--;
+}
+
+void Logic::inventoryAdd(int item) {
+	if (item >= 4)
+		_inventory[item] = 1;
+	else
+		_inventory[item] = 3;
+}
+
+bool Logic::wound() {
+	bool retval = false;
+	if (_health == 12 || _health == 8 || _health == 4) {
+		_health -= 2;
+		retval = true;
+	}
+	return retval;
+}
+
+bool Logic::chkDesertObstacles() {
+	if (!_vm->_wizard)
+		return false;
+
+	_nextNode = _nodeNum;
+	if (_lastObs) {
+		_lastObs = false;
+		return false;
+	}
+
+	if (_nodes[28] || _vm->_rnd->getRandomNumber(6) == 0) {
+		if (!_nodes[48] || _robberyNode != _nodeNum) {
+			if (_lastObstacle != _nodeNum) {
+				if (_vm->_rnd->getRandomNumber(5) == 0) {
+					_statPlay = 250;
+					_lastObstacle = _nodeNum;
+					_lastObs = true;
+					_vm->_loopFlag = true;
+					return true;
+				} else {
+					return false;
+				}
+			} else {
+				return false;
+			}
+		} else {
+			_statPlay = 490;
+			_vm->_loopFlag = true;
+			return true;
+		}
+	} else {
+		_statPlay = 280;
+		_robberyNode = _nodeNum;
+		_lastObstacle = _nodeNum;
+		_lastObs = true;
+		_vm->_loopFlag = true;
+		return true;
+	}
+}
+
+void Logic::increaseHealth() {
+	if (_health <= 3)
+		_health = 4;
+	else if (_health <= 7)
+		_health = 8;
+	else
+		_health = 12;
+}
+
+void Logic::endCredits() {
+	_nodeNum = 0;
+	_vm->_bTimer = 190;
+	_vm->readMouse();
+
+	while(_vm->_bTimer != 0 && _vm->_mouseButton == 0) {
+		_vm->checkTimers();
+		_vm->refreshSound();
+		_vm->checkMainScreen();
+		_vm->readMouse();
+	}
+
+	_vm->fadeToBlack1();
+	_vm->drawRect(4, 17, 228, 161, 0);
+	_vm->playSound(0);
+	dsAll();
+	_vm->_iconsClosed = true;
+	_vm->playMovie(199);
+	dsAll();
+	_vm->playMovie(205);
+	dsAll();
+	_vm->fadeToBlack1();
+	_vm->drawRect(4, 17, 228, 161, 0);
+	_vm->playMovie(201);
+	_vm->fShowPic(125);
+	_currMap = 3;
+	dsAll();
+	_pouch = false;
+	memset(_inventory, 0xFF, 19);
+	_statPlay = 994;
+	_vm->_loopFlag = true;
+}
+
+void Logic::gameHelp_Sub43C() {
+	_vm->fadeToBlack1();
+	_currMap = _vm->_asMap;
+	_vm->drawRect(4, 17, 228, 161, 0);
+	_vm->restoreAS();
+	_vm->_userInput = 0;
+	_vm->_gameMode = 0;
+	_vm->_iconsClosed = false;
+	_vm->_treeLeftSta = _vm->_oldTLS;
+	_eye = _oldEye;
+	_help = _oldHelp;
+	_pouch = _oldPouch;
+}
+
+void Logic::gameHelp() {
+	if (!_vm->_gameMode) {
+		if (_vm->_userInput == 0x43C) {
```

Comment:
Again, please add an enum of these hardcoded values

Reply:
the value of userInput is defined in the hotspots. Some values (like 0x190) are used in A LOT of scenes with different meanings, I don't know exactly how to handle that properly